sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java

```diff
@@ -0,0 +1,595 @@
+      String endPoint = String.format("/%s/%s/_search",
+                                      source.spec.getConnectionConfiguration().getIndex(),
+                                      source.spec.getConnectionConfiguration().getType());
+      Map<String, String> params = new HashMap<>();
+      params.put("scroll", source.spec.getScrollKeepalive() != null ? source.spec
+          .getScrollKeepalive() : "1m");
+      params.put("size", "1");
+      if (source.shardPreference != null) {
+        params.put("preference", "_shards:" + source.shardPreference);
+      }
+      HttpEntity queryEntity = new NStringEntity(query, ContentType.APPLICATION_JSON);
+      response =
+          restClient.performRequest("GET", endPoint, params, queryEntity, new BasicHeader("", ""));
+      JsonObject searchResult = parseResponse(response);
+      scroll = searchResult.getAsJsonPrimitive("_scroll_id").getAsString();
+      return updateCurrent(searchResult);
+    }
+
+    @Override
+    public boolean advance() throws IOException {
+      String requestBody = String.format("{\"scroll\" : \"%s\",\"scroll_id\" : \"%s\"}",
+                                         source.spec.getScrollKeepalive() != null ? source.spec
+                                             .getScrollKeepalive() : "1m", scroll);
+      HttpEntity scrollEntity = new NStringEntity(requestBody, ContentType.APPLICATION_JSON);
+      Response response = restClient.performRequest("GET", "/_search/scroll",
+                                                    Collections.<String, String>emptyMap(),
+                                                    scrollEntity,
+                                                    new BasicHeader("", ""));
+      JsonObject searchResult = parseResponse(response);
+      return updateCurrent(searchResult);
+    }
+
+    private boolean updateCurrent(JsonObject searchResult) {
+      //stop if no more data
+      if (searchResult.getAsJsonObject("hits").getAsJsonArray("hits").size() == 0) {
+        return false;
+      }
+      current = searchResult.getAsJsonObject("hits").getAsJsonArray("hits").get(0)
+          .getAsJsonObject().getAsJsonObject("_source").toString();
+      return true;
+    }
+
+    @Override
+    public String getCurrent() throws NoSuchElementException {
+      return current;
+    }
+
+    @Override
+    public void close() throws IOException {
+      // remove the scroll
+      String requestBody = String.format("{\"scroll_id\" : [\"%s\"]}", scroll);
+      HttpEntity entity = new NStringEntity(requestBody,
+                                            ContentType.APPLICATION_JSON);
+      try {
+        restClient.performRequest("DELETE", "/_search/scroll",
+                                  Collections.<String, String>emptyMap(),
+                                  entity, new BasicHeader("", ""));
+      } catch (IOException e) {
+        //ES gives the same scroll id to each source, close that scroll more than once produces an
```

Comment:
I don't quite understand this comment. Same scroll id across shards, or across something else?
Is there a piece of docs you can link to here?